### ISSUE TITLE
refetchEvents for multiple specific sources

### ISSUE THREAD

**Issue by benoitmasson** [fullcalendar/fullcalendar#254]
Hi,

I'm using both regular and background events on my calendar. Background events are used to highlight forbidden places for regular events, and are updated dynamically whenever such an event is selected (`eventClick()` or `eventDragStart()` callbacks) via the method '`refetchEvents`'.

In the case of a click, everything works well. In the case of a drag however, the '`refetchEvents`' causes the selected event to be refetched as well, thus leading to graphical glitches (event displayed twice, initial event not moving, ...).

The best workaround I found for this issue is to set a global property to prevent that the event is displayed (in the `eventRender()` callback) while it is dragged, and then in the `eventDrop()` callback I can manually update its start and end times according to delta. This is a ugly hack and it doesn't solve all bugs in my application.

A much simpler solution to that would be to allow to refetch events from a single source, leaving other events unchanged, but I don't think this is possible with the current API.

Is there something I am missing?

Thanks in advance

**Comment by benoitmasson** [comment]
In case anybody else needs, I made a patch (pull request above) to add a 'refetchBgEvents' method, which will refetch background events only and leave other events unchanged.

**Comment by arshaw** [comment]
should come up with an API that handles any criteria ("discussing" until we figure that out)

**Comment by benoit-masson** [comment]
Thanks Adam for the updates.

I agree, my pull request is very specific, but that was easy to code and sufficient for my needs.

As a more generic extension, I can propose the following:
* `refetchEvents(sourceFilter)` would refetch events only for sources that match the filter `sourceFilter`.

  A filter `sourceFilter` is an object, with attributes taken from the ones of event source objects  ([Documentation](http://fullcalendar.io/docs/event_data/Event_Source_Object/)). For example, if `sourceFilter = { color: 'yellow' }`, all sources which display events in yellow would be refetched.

  As a particular case of that, `refetchEvents(source)` would refetch only the source object `source`.

* `refetchEvents(sourceFilters)` or `refetchEvents(sourceFilter1, sourceFilter2, ...)`, where `sourceFilters = [sourceFilter1, sourceFilter2, ...]` would refetch sources that match any of the source filters.

  For example,
  ```javascript
refetchEvents([
        { rendering: background },
        { rendering: inverse-background }
]);
```
  would do exactly the same as my `refetchBgEvents()` method (refetch all events from all sources that render background or inverse background events)

* Finally, `refetchEvents()` would refetch all sources matching the empty set of filters, i.e., all sources (default behavior unchanged).


What do you think of that? It that's suitable, I can try to implement it (with no guarantee of success however, from what I've seen, this will be harder to integrate while keeping the code clean).

**Comment by benoitmasson** [comment]
Thanks Adam for the updates.

I agree, my pull request is very specific, but that was easy to code and sufficient for my needs.

As a more generic extension, I can propose the following:
* `refetchEvents(sourceFilter)` would refetch events only for sources that match the filter `sourceFilter`.

  A filter `sourceFilter` is an object, with attributes taken from the ones of event source objects  ([Documentation](http://fullcalendar.io/docs/event_data/Event_Source_Object/)). For example, if `sourceFilter = { color: 'yellow' }`, all sources which display events in yellow would be refetched.

  As a particular case of that, `refetchEvents(source)` would refetch only the source object `source`.

* `refetchEvents(sourceFilters)` or `refetchEvents(sourceFilter1, sourceFilter2, ...)`, where `sourceFilters = [sourceFilter1, sourceFilter2, ...]` would refetch sources that match any of the source filters.

  For example,
  ```javascript
refetchEvents([
        { rendering: background },
        { rendering: inverse-background }
]);
```
  would do exactly the same as my `refetchBgEvents()` method (refetch all events from all sources that render background or inverse background events)

* Finally, `refetchEvents()` would refetch all sources matching the empty set of filters, i.e., all sources (default behavior unchanged).


What do you think of that? It that's suitable, I can try to implement it (with no guarantee of success however, from what I've seen, this will be harder to integrate while keeping the code clean).

**Comment by arshaw** [comment]
this looks good :+1:

**Comment by arshaw** [comment]
of course very related to https://github.com/fullcalendar/fullcalendar/issues/1328

**Comment by arshaw** [comment]
per our discussion in #3103, this has been renamed to `fetchEventSources`

**Comment by arshaw** [comment]
[Code complete](https://github.com/fullcalendar/fullcalendar/commits/v2.8.0-dev) and slated for v2.8.0 ([wip docs](https://github.com/fullcalendar/fullcalendar/commits/docs-2.0)). Will notify when released.

**Comment by arshaw** [comment]
Released in [v2.8.0](https://github.com/fullcalendar/fullcalendar/releases/tag/v2.8.0).

http://fullcalendar.io/docs/event_data/refetchEventSources/
(called with a single source argument)

**Comment by benoitmasson** [comment]
Great job, thanks @caseyjhol for the hard work!
